proj-xcode/Classes/keychain/PA2Keychain.m

```diff
@@ -264,33 +264,112 @@ - (NSDictionary*) allItemsWithPrompt:(NSString*)prompt withStatus: (OSStatus *)s
 
 #pragma mark - Biometry support
 
-+ (BOOL) canUseBiometricAuthentication
-{
-	return [self supportedBiometricAuthentication] != PA2SupportedBiometricAuthentication_None;
-}
+/**
+ Looks like Apple introduced a new biometry type. We should try to continue,
+ and pretend that TouchID is available. Application's UI will probably display
+ wrong information, but at least it may work.
+ */
+#define __BiometricAuthenticationType_Fallback 	PA2BiometricAuthenticationType_TouchID
 
-+ (PA2SupportedBiometricAuthentication) supportedBiometricAuthentication
+/**
+ A private function returns full information about biometric support on the system. The method internally
+ uses `LAContext.canEvaluatePolicy()`.
+ */
+static PA2BiometricAuthenticationInfo _getBiometryInfo()
 {
+	PA2BiometricAuthenticationInfo info = { PA2BiometricAuthenticationStatus_NotSupported, PA2BiometricAuthenticationType_None };
 #if !defined(PA2_EXTENSION_SDK)
 	// Check is available only for regular iOS applications
 	if (@available(iOS 9, *)) {
 		LAContext * context = [[LAContext alloc] init];
-		BOOL canEvaluate = [context canEvaluatePolicy:kLAPolicyDeviceOwnerAuthenticationWithBiometrics error:nil];
+		NSError * error = nil;
+		BOOL canEvaluate = [context canEvaluatePolicy:kLAPolicyDeviceOwnerAuthenticationWithBiometrics error:&error];
 		if (canEvaluate) {
+			// If we can evaluate, then everything is quite simple.
+			info.currentStatus = PA2BiometricAuthenticationStatus_Available;
+			// Now check the type of biometry
 			if (@available(iOS 11.0, *)) {
 				LABiometryType bt = context.biometryType;
 				if (bt == LABiometryTypeTouchID) {
-					return PA2SupportedBiometricAuthentication_TouchID;
+					info.biometryType = PA2BiometricAuthenticationType_TouchID;
 				} else if (bt == LABiometryTypeFaceID) {
-					return PA2SupportedBiometricAuthentication_FaceID;
+					info.biometryType = PA2BiometricAuthenticationType_FaceID;
+				} else {
+					// Check comment above "__BiometricAuthenticationType_Fallback" for details.
+					PA2Log(@"Warning: LAContext.biometryType contains unknown biometryType %@.", @(bt));
+					info.biometryType = __BiometricAuthenticationType_Fallback;
+				}
+			} else {
+				// No FaceID before iOS11, so it has to be TouchID
+				info.biometryType = PA2BiometricAuthenticationType_TouchID;
+			}
+			//
+		} else {
+			// In case of error we cannot evaluate, but the type of biometry can be determined.
+			NSInteger code = [error.domain isEqualToString:LAErrorDomain] ? error.code : 0;
+			if (@available(iOS 11.0, *)) {
+				// On iOS 11 its quite simple, we have type property available and status can be determied
+				// from the error.
+				LABiometryType bt = context.biometryType;
```

Comment:
This seems to be the same code as the one in case of no error. Could we simplify the code (use `context.biometryType` for type of biometry and `canEvaluate`/error for status)?

Reply:
I was thinking about that, but in my opinion, it will lead to more complex and less readable code. That @availability check has its internal cost, so I was trying to reduce its usage to minimum. Current impl. requires only one check for IOS9 and another for IOS11 for both, error & success branch. 

But I can move that enumeration translation to another function, which will be probably inlined.

Reply:
OK, makes sense - let's leave the code as it is, the optimization (if any) would be marginal...

Reply:
My last commit moved that repeated code block to translation function. It's much better I think.